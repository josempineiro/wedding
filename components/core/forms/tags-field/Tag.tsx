import classNames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faMinus } from "@fortawesome/free-solid-svg-icons";
import { Button } from "@/components/core/buttons/Button";

export interface TagProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
  onClose?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  closable?: boolean;
}

export function Tag({
  children,
  className,
  onClose,
  closable,
  ...rest
}: TagProps) {
  return (
    <div
      role="tag"
      className={classNames([
        className,
        "inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white",
        {
          "cursor-default": !closable,
          "cursor-pointer": closable,
          "pr-0": closable,
        },
      ])}
      {...rest}
    >
      <span className="text-sm">{children}</span>
      {closable && (
        <Button onClick={onClose} size="sm" variant="primary" role="delete-tag">
          <FontAwesomeIcon icon={faMinus} />
        </Button>
      )}
    </div>
  );
}
