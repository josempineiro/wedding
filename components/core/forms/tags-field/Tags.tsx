import React, { useState } from "react";
import classNames from "classnames";
import { Tag } from "@/components/core/forms/tags-field/Tag";

export type TagsProps = React.HTMLAttributes<HTMLDivElement> & {
  tags: string[];
  maxTagsToShow?: number;
  closable?: boolean;
  onClose?: (tag: string) => void;
} & (
    | {
        closable: true;
        onClose: (tag: string) => void;
      }
    | {
        closable?: false;
      }
  );

export function Tags({
  tags,
  className,
  maxTagsToShow = Infinity,
  closable,
  onClose,
  ...rest
}: TagsProps) {
  const [showAllTags, setShowAllTags] = useState(false);
  const visibleTags = showAllTags ? tags : tags.slice(0, maxTagsToShow);
  const hiddenTags = tags.slice(maxTagsToShow);
  const handleClickMoreTags = () => {
    setShowAllTags(!showAllTags);
  };
  const handleCloseTag = (tag: string) => (event: React.MouseEvent) => {
    event.stopPropagation();
    if (closable) {
      onClose(tag);
    }
  };
  return (
    <div
      {...rest}
      className={classNames([className, "flex flex-wrap gap-2 items-center"])}
    >
      {visibleTags.map((tag) => (
        <Tag key={tag} closable={closable} onClose={handleCloseTag(tag)}>
          {tag}
        </Tag>
      ))}
      {hiddenTags.length > 0 && (
        <Tag role="more-tags" onClick={handleClickMoreTags}>
          <>{showAllTags ? "Hide" : `+${hiddenTags.length} more`} </>
        </Tag>
      )}
    </div>
  );
}
