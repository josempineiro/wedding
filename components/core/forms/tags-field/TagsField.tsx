import { useState } from "react";
import { FieldType } from "@/domain/core/forms/fields/Field";
import { Tags } from "@/components/core/forms/tags-field/Tags";
import { Field, type FieldProps } from "@/components/core/forms/field/Field";
import { Input, type InputProps } from "@/components/core/forms/input/Input";

type TagsType = FieldType & Array<string>;
export interface TagsFieldProps
  extends Omit<InputProps<string>, "onChange" | "value"> {
  label?: FieldProps<string[]>["label"];
  type?: "text" | "email" | "password";
  errors?: FieldProps<string[]>["errors"];
  inputRef?: React.Ref<HTMLInputElement>;
  value: string[];
  onChange?: (value: string[]) => void;
  onAdd?: (value: string) => void;
  name: FieldProps<string[]>["name"];
  maxTagsToShow?: number;
}

export function TagsField({
  label,
  errors,
  name,
  inputRef,
  onChange = () => {},
  onAdd = () => {},
  disabled,
  readOnly,
  value,
  maxTagsToShow,
  ...rest
}: TagsFieldProps) {
  const [tagInputValue, setTagInputValue] = useState<string>("");
  const handleAdd = (tag: string) => {
    onAdd(tag);
    onChange([...value, tag]);
    setTagInputValue("");
  };

  const handleDelete = (deletedTag: string) => {
    onChange(value.filter((tag) => tag !== deletedTag));
  };
  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (
      event.key === "Enter" &&
      (event.target as HTMLInputElement).value !== ""
    ) {
      event.preventDefault();
      handleAdd(event.currentTarget.value);
    }
  };
  const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
    if ((event.target as HTMLInputElement).value !== "") {
      event.preventDefault();
      handleAdd(event.currentTarget.value);
    }
  };
  return (
    <Field type="tags" errors={errors} label={label} name={name}>
      <div className="flex bg-white">
        {value.length > 0 && (
          <Tags
            tags={value}
            onClose={handleDelete}
            closable={!disabled && !readOnly}
            maxTagsToShow={maxTagsToShow}
            className="bg-white px-2"
          />
        )}
        {!disabled && !readOnly && (
          <Input<string>
            ref={inputRef}
            name={name}
            value={tagInputValue}
            className="bg-white px-2 py-1"
            onKeyDown={handleKeyDown}
            onChange={setTagInputValue}
            onBlur={handleBlur}
            {...rest}
          />
        )}
      </div>
    </Field>
  );
}
