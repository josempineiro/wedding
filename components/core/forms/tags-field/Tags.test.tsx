import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Tags } from "@/components/core/forms/tags-field/Tags";

describe("Tags component", () => {
  const tags = ["Tag 1", "Tag 2", "Tag 3"];
  const onClose = jest.fn();

  it("should render visible tags", () => {
    const { container } = render(<Tags tags={tags} maxTagsToShow={2} />);

    const visibleTags = screen.getAllByRole("tag");

    expect(visibleTags).toHaveLength(2);
    expect(visibleTags[0]).toHaveTextContent("Tag 1");
    expect(visibleTags[1]).toHaveTextContent("Tag 2");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render hidden tags when click in more tags", () => {
    const { container } = render(
      <Tags tags={tags} maxTagsToShow={2} onClose={onClose} />
    );

    const moreTag = screen.getByRole("more-tags");

    fireEvent.click(moreTag);

    const visibleTags = screen.getAllByRole("tag");

    expect(visibleTags).toHaveLength(3);
    expect(visibleTags[0]).toHaveTextContent("Tag 1");
    expect(visibleTags[1]).toHaveTextContent("Tag 2");
    expect(visibleTags[2]).toHaveTextContent("Tag 3");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call onClose when closing a tag", () => {
    render(<Tags tags={tags} closable onClose={onClose} />);

    const closeButtons = screen.getAllByRole("delete-tag");

    fireEvent.click(closeButtons[0]);

    expect(onClose).toHaveBeenCalledWith("Tag 1");
  });
});
