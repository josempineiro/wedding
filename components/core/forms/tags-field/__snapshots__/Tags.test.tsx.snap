// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags component should render hidden tags when click in more tags 1`] = `
<div
  class="flex flex-wrap gap-2 items-center"
>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="tag"
  >
    <span
      class="text-sm"
    >
      Tag 1
    </span>
  </div>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="tag"
  >
    <span
      class="text-sm"
    >
      Tag 2
    </span>
  </div>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="tag"
  >
    <span
      class="text-sm"
    >
      Tag 3
    </span>
  </div>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="more-tags"
  >
    <span
      class="text-sm"
    >
      Hide
       
    </span>
  </div>
</div>
`;

exports[`Tags component should render visible tags 1`] = `
<div
  class="flex flex-wrap gap-2 items-center"
>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="tag"
  >
    <span
      class="text-sm"
    >
      Tag 1
    </span>
  </div>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="tag"
  >
    <span
      class="text-sm"
    >
      Tag 2
    </span>
  </div>
  <div
    class="inline-flex items-center gap-2 px-2 rounded-sm bg-primary-400 text-white cursor-default"
    role="more-tags"
  >
    <span
      class="text-sm"
    >
      +1 more
       
    </span>
  </div>
</div>
`;
